### PR TITLE
[inductor] Move ProxyExecutor behind the C shim layer

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109673
* #109606
* #109450
* __->__ #109498
* #109436

Summary: Move ProxyExecutor behind the C shim layer for ABI compatibility

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/csrc/inductor/aoti_torch/shim_common.cpp [hunk 1]
```diff
@@ -71,32 +74,28 @@ int32_t aoti_torch_dtype_int64() {
 
 AOTITorchError aoti_torch_delete_tensor_object(AtenTensorHandle tensor) {
   AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
-    at::Tensor* t =
-        torch::aot_inductor::tensor_handle_to_tensor_pointer(tensor);
+    at::Tensor* t = tensor_handle_to_tensor_pointer(tensor);
```
Comment (ezyang) [inline torch/csrc/inductor/aoti_torch/shim_common.cpp:77]: This refactor is totally unrelated

**Review by ezyang — APPROVED** [review]

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Comment by desertfire** [comment]
Landed with an exported combined PR, https://github.com/pytorch/pytorch/pull/109678
